traffic_control/clients/python/to_access/__init__.py

```diff
@@ -0,0 +1,620 @@
+
+	try:
+		to_host = args.to_host if args.to_host else os.environ["TO_URL"]
+	except KeyError as e:
+		raise KeyError("Traffic Ops hostname not set! Set the TO_URL environment variable or use "\
+		               "'--to_url'.") from e
+
+	useSSL = True
+	to_port = 443
+	if to_host.lower().startswith("http://"):
+		to_host = to_host[7:]
+		useSSL = False
+		to_port = 80
+	elif to_host.lower().startswith("https://"):
+		to_host = to_host[8:]
+
+	portpos = to_host.find(':')
+	if portpos > 0:
+		to_port = int(to_host[portpos+1:])
+		to_host = to_host[:portpos]
+
+	s = TOSession(to_host,
+	              host_port=to_port,
+	              ssl=useSSL,
+	              api_version=str(args.api_version),
+	              verify_cert=not args.insecure)
+	# s.create()
+
+	# The TOSession object's methods will handle '/' stripping for us, so it's only necessary with
+	# raw paths.
+	path = args.PATH if not args.raw_path else '/'.join((s.to_url.rstrip('/'), args.PATH.lstrip('/')))
+
+	data = args.DATA
+	if data and os.path.isfile(data):
+		with open(data) as f:
+			data = f.read()
+
+	if isinstance(data, str):
+		data = data.encode()
+
+	# if os.path.isfile(COOKIEFILE):
+	# 	with open(COOKIEFILE) as f:
+	# 		s._session.cookies.set("mojolicious", f.read())
+	# 	return s, path, data, args.full, args.raw_path
+
+	try:
+		to_user = args.to_user if args.to_user else os.environ["TO_USER"]
+	except KeyError as e:
+		raise KeyError("Traffic Ops user not set! Set the TO_USER environment variable or use "\
+		               "'--to_user'.") from e
+
+	try:
+		to_passwd = args.to_password if args.to_password else os.environ["TO_PASSWORD"]
+	except KeyError as e:
+		raise KeyError("Traffic Ops password not set! Set the TO_PASSWORD environment variable or "\
+		               "use '--to_password'") from e
+
+	try:
+		s.login(to_user, to_passwd)
+	except (OperationError, InvalidJSONError, LoginError) as e:
+		raise PermissionError from e
+
+	return (s,
+	       path,
+	       data,
+	       (
+	         args.request_headers or args.full,
+	         args.response_headers or args.full,
+	         args.request_payload or args.full
+	       ),
+	       args.raw_path,
+	       args.pretty)
+
+
+def get():
+	"""
+	Entry point for :program:`toget`
+
+	:returns: The program's exit code
+	"""
+	try:
+		s, path, data, full, raw, pretty = parse_arguments("toget")
+	except (PermissionError, KeyError) as e:
+		print(e, file=sys.stderr)
+		return 1
+
+	try:
+		if raw:
+			if data is not None:
+				r = s._session.get(path, data=data)
+			else:
+				r = s._session.get(path)
+		elif data is not None:
+			r = s.get(path, data=data)[1]
+		else:
+			r = s.get(path)[1]
+	except (OperationError, InvalidJSONError) as e:
+		if e.resp is not None:
+			r = e.resp
+		else:
+			print("Error occurred: ", e, file=sys.stderr)
+			return 2
+
+	output(r, pretty, *full)
+	# set_cookie(s._session.cookies.get("mojolicious"))
+	return 0 if r.status_code < 400 else r.status_code // 100
+
+def put():
```

Comment:
These functions for the different HTTP methods are really duplicated other than the argument to `parse_arguments` and the `s._session.<method>(...)`. Can we collapse them down into a shared function that takes those as parameters?

Reply:
yes and no. They need to be separate functions so that the scripts produced by `setuptools` can have separate entrypoints. They can, of course, be made into stubs that call a single shared function, but that involves _always_ using the `_session` member of the `TOSession` object which is "private" (although I'm all for obliterating that). Not a big deal, really, because I already have to do that for "raw" request paths, but it's a bit of extra work to build the URL since I won't be getting `/api/{{version}}/` prepended for free.

Reply:
Yeah, I think separate functions are fine, just that they would all use the same shared function that varies on the  `parse_arguments` param and `session.<method>`, however you'd like to handle that.

Reply:
Done